nucliadb_procs/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2021 Bosutech XXI S.L.
+//
+// nucliadb is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at info@nuclia.com.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, ItemFn, LitStr, Token};
+
+/// Measure time elapsed during the wrapped function call.
+///
+/// It records a `nucliadb_core` request time metric, so `nucliadb_core` *must*
+/// be available when using this macro.
+#[proc_macro_attribute]
+pub fn measure(
+    args: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    println!("Args: {}", args);
+    println!("Item: {}", item);
+
+    let MeasureArgs { actor, metric } = parse_macro_input!(args as MeasureArgs);
+
+    let ItemFn {
+        attrs,
+        vis,
+        sig,
+        block,
+    } = parse_macro_input!(item as ItemFn);
+
+    let actor = format_ident!("{}", actor);
+    let metric = match metric {
+        Some(s) => s,
+        None => sig.ident.to_string(),
+    };
+
+    let expanded = quote! {
+        #(#attrs)*
+        #vis #sig {
+            use std::time::SystemTime;
+            use nucliadb_core::metrics;
+            use nucliadb_core::metrics::request_time::RequestTimeKey;
+
+            let time = SystemTime::now();
+
+            // execute function body
+            let return_value = #block;
+
+            let took = time.elapsed().map(|elapsed| elapsed.as_secs_f64()).unwrap_or(f64::NAN);
```

Comment:
it would be nice to be able to pass to the macro the desired unit. micro second, milliseconds or seconds

Reply:
oh mm maybe prometheus takes only seconds in the current code, so nevermind. that said, I would defintiely use that macro in our benchmarks without sending the metrics just to get the value.

Reply:
We have different implementation of meters (trait to collect metrics). As the macro is using `metrics::get_metrics` we only need to set the desired meter and it'll printed instead of sent to Prometheus, for example (`ConsoleMeter` instead of `PrometheusMeter`)